mesonbuild/interpreter/compiler.py

```diff
@@ -797,3 +806,54 @@ def preprocess_method(self, args: T.Tuple[T.List['mesonlib.FileOrString']], kwar
         # other targets, list outputs, etc.
         private_dir = os.path.relpath(self.interpreter.backend.get_target_private_dir(tg), self.interpreter.subdir)
         return [build.CustomTargetIndex(tg, os.path.join(private_dir, o)) for o in tg.outputs]
+
+    @typed_kwargs('compiler.symbol_visibility_header',
+                  KwargInfo('api', (str, NoneType)),
+                  KwargInfo('compilation', (str, NoneType)),
+                  KwargInfo('static_compilation', (str, NoneType)),
+                  KwargInfo('static_only', bool, default=False))
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '1.1.0')
+    def symbol_visibility_header_method(self, args: T.Tuple[str], kwargs: 'SymbolVisibilityHeaderKW') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp', 'objc', 'objcpp'}:
+            raise InterpreterException('compiler.symbol_visibility_header is supported only for C and C++ compilers')
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        if default_prefix[0].isdigit():
+            default_prefix = f'_{default_prefix}'
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = '#pragma once\n'
+        if static_only:
+            content += textwrap.dedent(f'''
+                #ifndef {static_compilation}
+                #  define {static_compilation}
+                #endif /* {static_compilation} */
+                ''')
+        content += textwrap.dedent(f'''
+            #if (defined(_WIN32) || defined(__CYGWIN__)) && !defined({static_compilation})
+            #  define _{api}_EXPORT __declspec(dllexport)
```

Comment:
Why are these prefixed with underscore? Underscore-prefixed macros are reserved for the compiler

Reply:
I though that the double leading underscore is reserved, not the single one.

Reply:
I think that as well, but if in doubt I can use triple underscore lol

Reply:
https://stackoverflow.com/a/17307796/7572728 has some good info

Reply:
Why do they even need to be private? :D

Reply:
> Why do they even need to be private? :D

Fair enough, removed the underscore. There is even a niche use-case for using `MYPROJECT_API_EXPORT`macro, glib does it for functions declarations that are exposed by modules: https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/giomodule.h#L130.